## **Collaborative Filtering (Item Based)**

In [0]:
#Importing relevant libraries

import numpy as np
import pandas as pd

In [0]:
#Loading csv files from drive

anime = pd.read_csv('/content/drive/My Drive/anime/anime.csv')
reviews = pd.read_csv('/content/drive/My Drive/anime/rating.csv')

## **Inspecting data**

In [20]:
#Looking at the contents in anime.csv file

anime.head()

anime_id                              name  ... rating members
0     32281                    Kimi no Na wa.  ...   9.37  200630
1      5114  Fullmetal Alchemist: Brotherhood  ...   9.26  793665
2     28977                          Gintama°  ...   9.25  114262
3      9253                       Steins;Gate  ...   9.17  673572
4      9969                     Gintama&#039;  ...   9.16  151266

[5 rows x 7 columns]

In [21]:
#Looking at the contents in ratings.csv file

reviews.head()

user_id  anime_id  rating
0        1        20      -1
1        1        24      -1
2        1        79      -1
3        1       226      -1
4        1       241      -1

In [7]:
reviews.tail()

user_id  anime_id  rating
7813732    73515     16512       7
7813733    73515     17187       9
7813734    73515     22145      10
7813735    73516       790       9
7813736    73516      8074       9

## **Data Cleaning**

In [22]:
#Checking for missing values in anime file

anime.isnull().sum()

anime_id      0
name          0
genre        62
type         25
episodes      0
rating      230
members       0
dtype: int64

####*62 Genres and 25 types are not specified*
####*230 ratings are not given!*

In [23]:
#Checking for missing values in ratings file

reviews.isnull().sum()

user_id     0
anime_id    0
rating      0
dtype: int64

*No missing values found..* 🎉🎊

In [24]:
#Dealing with the missing values

anime['type'] = anime['type'].fillna('None')
anime['genre'] = anime['genre'].fillna('None')
anime['rating'] = anime['rating'].fillna(anime['rating'].median())
anime.isnull().sum()

anime_id    0
name        0
genre       0
type        0
episodes    0
rating      0
members     0
dtype: int64

In [26]:
len(reviews) #Total length of ratings table

7813737

In [25]:
len(anime) #Total length of anime table

12294

## **Using Cosine Similarity**

In [0]:
listOfDictonaries=[]
indexMap = {}
reverseIndexMap = {}
ptr=0;
test = reviews
test=test[['user_id','rating']].groupby(test['anime_id']) 


for groupKey in test.groups.keys():
    tempDict={}

    group = test.get_group(groupKey)   

    for i in range(0,len(group)):
        tempDict[group.iloc[i,0]]=group.iloc[i,1]
    indexMap[ptr]=groupKey                
    reverseIndexMap[groupKey] = ptr       
    ptr=ptr+1
    listOfDictonaries.append(tempDict)

In [0]:
from sklearn.feature_extraction import DictVectorizer
dictVectorizer = DictVectorizer(sparse=True)
vector = dictVectorizer.fit_transform(listOfDictonaries)

In [15]:
from sklearn.metrics.pairwise import cosine_similarity
pairwiseSimilarity = cosine_similarity(vector)
print(pairwiseSimilarity)

[[1.00000000e+00 6.03567084e-01 5.30746446e-01 ... 1.12684709e-02
  9.62509381e-04 9.62509381e-04]
 [6.03567084e-01 1.00000000e+00 4.09950884e-01 ... 5.79678828e-03
  1.54043346e-03 1.54043346e-03]
 [5.30746446e-01 4.09950884e-01 1.00000000e+00 ... 7.70190515e-03
  1.21186040e-03 1.21186040e-03]
 ...
 [1.12684709e-02 5.79678828e-03 7.70190515e-03 ... 1.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [9.62509381e-04 1.54043346e-03 1.21186040e-03 ... 0.00000000e+00
  1.00000000e+00 1.00000000e+00]
 [9.62509381e-04 1.54043346e-03 1.21186040e-03 ... 0.00000000e+00
  1.00000000e+00 1.00000000e+00]]


In [0]:
def animedetails(anime_id):
    print("Name:", anime[anime['anime_id']==anime_id]['name'].values[0])
    print("Genres:",anime[anime['anime_id']==anime_id]['genre'].values[0])
    print("Anime-ID:",anime_id)
    print("-"*40)


def topRecommandations(anime_id):
    row = reverseIndexMap[anime_id]
    print("------Anime entered--------")
    animedetails(anime_id)
    print("-------Recommendations----------")

    similaranime = [animedetails(indexMap[i]) for i in np.argsort(pairwiseSimilarity[row])[-7:-2][::-1]]

In [17]:
topRecommandations(8)

------Anime entered--------
Name: Beet the Vandel Buster
Genres: Adventure, Fantasy, Shounen, Supernatural
Anime-ID: 8
----------------------------------------
-------Recommendations----------
Name: Onmyou Taisenki
Genres: Adventure, Fantasy, Shounen, Supernatural
Anime-ID: 613
----------------------------------------
Name: Kiba
Genres: Adventure, Fantasy, Shounen
Anime-ID: 845
----------------------------------------
Name: Rekka no Honoo
Genres: Action, Adventure, Martial Arts, Shounen, Super Power
Anime-ID: 238
----------------------------------------
Name: The Law of Ueki
Genres: Adventure, Comedy, Drama, Shounen, Supernatural
Anime-ID: 479
----------------------------------------
Name: Groove Adventure Rave
Genres: Adventure, Comedy, Fantasy, Romance, Shounen
Anime-ID: 246
----------------------------------------
